### PR TITLE
UTY-1082 (formerly UTY-1192) Component updated per field

### PR DESCRIPTION
**Pull request by Zomtorg**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Introduce a bitmask so that only fields that have been changed of a component are updated.
#### Tests
Manual test in playground. Unit tests for components with no fields.
#### Documentation
https://github.com/spatialos/gdk-for-unity/pull/583
#### Primary reviewers
@samiwh @jamiebrynes7

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 1]
```diff
@@ -19,7 +20,40 @@ namespace <#= qualifiedNamespace #>
         {
             public uint ComponentId => <#= componentDetails.ComponentId #>;
 
-            public BlittableBool DirtyBit { get; set; }
+            [Flags]
+            public enum DirtyBitsEnum
+            {
+                None = 0,
+<# int currentFieldBinaryIdx = 1; #>
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: If you change to using a normal for loop, you can use `i +1` instead of having a seperate variable

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 2]
```diff
@@ -19,7 +20,40 @@ namespace <#= qualifiedNamespace #>
...
+                <#= fieldDetails.PascalCaseName #> = <#= currentFieldBinaryIdx #>,
+<# currentFieldBinaryIdx = currentFieldBinaryIdx << 1; #>
+<# } #>
+            }
+
+            private DirtyBitsEnum dirtyBits;
+
+            public DirtyBitsEnum DirtyBits => dirtyBits;
+
+            private void AddDirtyBit(DirtyBitsEnum currentField)
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: These functions will be called a lot, but only from within generated code. Will be worth inlining them! Its generated code anyways, no harm in having it be less readable but faster.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 3]
```diff
@@ -19,7 +20,40 @@ namespace <#= qualifiedNamespace #>
...
+            {
+                dirtyBits = dirtyBits | currentField;
+            }
+
+            public bool ContainsDirtyBits()
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: `ContainsDirtyBits` -> `IsDirty`

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 2 above)
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: `AddDirtyBit` -> `MarkDirtyBit`
Since you arn't adding anything, just setting/marking in a bitmask

Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: (But I'd still prefer this one to be inlined)

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 1 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Done

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 2 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: That makes sense - I've inlined them now.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt [hunk 4]
```diff
@@ -542,7 +542,7 @@ namespace <#= qualifiedNamespace #>
                     dirtyEvents += events<#= eventDetail.EventName #>.Count;
 <# } #>
 
-                    if (data.DirtyBit || dirtyEvents > 0)
+                    if (data.IsDirty() || dirtyEvents > 0)
                     {
                         var update = new global::Improbable.Worker.Core.SchemaComponentUpdate(<#= componentDetails.ComponentId #>);
                         <#= componentNamespace #>.Serialization.SerializeUpdate(data, update);
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: This can now be skipped if `!data.IsDirty()` It'll save a lot of checking on components that just had events updated

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 5]
```diff
@@ -19,7 +20,33 @@ namespace <#= qualifiedNamespace #>
...
+            {
+                return (DirtyBits != DirtyBitsEnum.None);
+            }
+
+            public bool IsDirty(DirtyBitsEnum currentField)
+            {
+                return (dirtyBits & currentField) != DirtyBitsEnum.None;
+            }
+
+            public void ClearDirtyBits()
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: We don't actually need to specify that we are doing this with a bit mask in the name. The API is that you mark something as no dirty. Bits are just an implementation that the caller doesn't need to know.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 6]
```diff
@@ -19,7 +20,33 @@ namespace <#= qualifiedNamespace #>
...
+            {
+                None = 0,
+<# for(int i = 0; i < fieldDetailsList.Count; i++) { #>
+                <#= fieldDetailsList[i].PascalCaseName #> = <#= 1 << i #>,
+<# } #>
+            }
+
+            private DirtyBitsEnum dirtyBits;
+
+            public DirtyBitsEnum DirtyBits => dirtyBits;
```
[the remark this answers is not in the document]
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Otherwise we need set functions. Although given that this often called fields are better than properties in IComponentData objects.

[the remark this answers is not in the document]
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Set functions would be better IMO, most users will not like to deal with bit operations

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 7]
```diff
@@ -94,9 +121,7 @@ namespace <#= qualifiedNamespace #>
 <# if (!fieldDetails.IsBlittable) { #>
                 component.<#= fieldDetails.CamelCaseName#>Handle = <#= qualifiedNamespace #>.<#= componentDetails.ComponentName #>.ReferenceTypeProviders.<#= fieldDetails.PascalCaseName #>Provider.Allocate(world);
 <# } #>
-                {
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Removal of the `{}` around these may have harmful effects (I remember some deserialization strings used local variables which may clash if there are multiple properties of the same type in the same component) - a fix would be to place the `{}` into those string generators.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 6 above)
[the remark this answers is not in the document]
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Discussed offline. We will just have one public field for simplicity for now.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 7 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Rolling this back, just to be safe.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs [hunk 8]
```diff
@@ -51,12 +51,23 @@ protected override void OnUpdate()
...
-                    TicksPerSecond = tickRate.PhysicsTicksPerRealSecond,
-                    DirtyBit = true
+                    TicksPerSecond = tickRate.PhysicsTicksPerRealSecond
                 };
 
-                if (!(TransformUtils.HasChanged(currentTransform.Location, transform.Location) ||
-                    TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation)))
+                var locationHasChanged = TransformUtils.HasChanged(currentTransform.Location, transform.Location);
+                var rotationHasChanged = TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation);
+                var velocityHasChanged = TransformUtils.HasChanged(currentTransform.Velocity, transform.Velocity);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: This is deliberately not changing velocity right now.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs [hunk 9]
```diff
@@ -51,12 +51,23 @@ protected override void OnUpdate()
...
+                var ticksPerSecondHasChanged =
+                    Mathf.Abs(currentTransform.TicksPerSecond - transform.TicksPerSecond) < Mathf.Epsilon;
+
+                currentTransform.DirtyBits |= locationHasChanged ? TransformInternal.Component.DirtyBitsFlag.Location : TransformInternal.Component.DirtyBitsFlag.None;
+                currentTransform.DirtyBits |= rotationHasChanged ? TransformInternal.Component.DirtyBitsFlag.Rotation : TransformInternal.Component.DirtyBitsFlag.None;
+                currentTransform.DirtyBits |= velocityHasChanged ? TransformInternal.Component.DirtyBitsFlag.Velocity : TransformInternal.Component.DirtyBitsFlag.None;
+                currentTransform.DirtyBits |= physicsTickHasChanged ? TransformInternal.Component.DirtyBitsFlag.PhysicsTick : TransformInternal.Component.DirtyBitsFlag.None;
+                currentTransform.DirtyBits |= ticksPerSecondHasChanged ? TransformInternal.Component.DirtyBitsFlag.TicksPerSecond : TransformInternal.Component.DirtyBitsFlag.None;
+
+                if (!locationHasChanged && rotationHasChanged)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: missing a !

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 5 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: MarkNotDirty would be better

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs [hunk 10]
```diff
@@ -51,12 +51,23 @@ protected override void OnUpdate()
...
-                if (!(TransformUtils.HasChanged(currentTransform.Location, transform.Location) ||
-                    TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation)))
+                var locationHasChanged = TransformUtils.HasChanged(currentTransform.Location, transform.Location);
+                var rotationHasChanged = TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation);
+                var velocityHasChanged = TransformUtils.HasChanged(currentTransform.Velocity, transform.Velocity);
+                var physicsTickHasChanged = currentTransform.PhysicsTick == transform.PhysicsTick;
+                var ticksPerSecondHasChanged =
+                    Mathf.Abs(currentTransform.TicksPerSecond - transform.TicksPerSecond) < Mathf.Epsilon;
+
+                currentTransform.DirtyBits |= locationHasChanged ? TransformInternal.Component.DirtyBitsFlag.Location : TransformInternal.Component.DirtyBitsFlag.None;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Would this not be done by setting the fields?

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 9 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: 😱 Fixed!

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt [hunk 11]
```diff
@@ -555,7 +555,7 @@ namespace <#= qualifiedNamespace #>
                         dirtyEvents += events<#= eventDetail.EventName #>.Count;
 <# } #>
 
-                        if (data.DirtyBit || dirtyEvents > 0)
+                        if (data.IsDirty() || dirtyEvents > 0)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: Drive-by. We should also change the name of dirty events.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 12]
```diff
@@ -24,7 +24,32 @@ namespace <#= qualifiedNamespace #>
...
             public uint ComponentId => <#= componentDetails.ComponentId #>;
 
-            public BlittableBool DirtyBit { get; set; }
+            // Represents a bit mask for tracking which component properties were changed locally and need to be synced.
+            [Flags]
+            public enum DirtyBitsFlag
+            {
+                None = 0,
+<# for(int i = 0; i < fieldDetailsList.Count; i++) { #>
+                <#= fieldDetailsList[i].PascalCaseName #> = <#= 1 << i #>,
```
[the remark this answers is not in the document]
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: I think we can do a field count check and either use an int32 or in64 enum? (the 2 supported options in C#)
Any component with more then 64 fields should fail to generate. I doubt limiting Components to only 64 fields would be a painful limitation?

[the remark this answers is not in the document]
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: I've asked production (https://improbable.slack.com/archives/C9PR5EA21/p1540377195000100) and it looks like nobody right now is using more than 64 fields. There are hypothetical use cases where you define elements of a fixed length collection as explicit fields to avoid using schema lists/collections but that use case should have a limited shelf life with delta compressed component updates around the corner. My suggestion would be: Let's go with @zeroZshadow's suggestion for now, add a known issue note in the docs and see how many people actually complain about this limitation. This might end up being a known issue that we'll never fix. @samiwh  are you on-board with this?

[the remark this answers is not in the document]
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Sounds fair. I will look into stackallocing a byte array.

[the remark this answers is not in the document]
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: stackalloc is not something that works here.
If we want beyond 64 fields, we'd need have to have multiple bit flag fields.
I still think a limitation of 64 fields is a totally valid restraint

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 8 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Reverted to the old logic.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 10 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: I've changed the code to set the dirty bits implicitly through the setter properties.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 5 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Done!

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt (hunk 11 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: Fixed!

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs [hunk 13]
```diff
@@ -45,18 +45,17 @@ protected override void OnUpdate()
...
-                var currentTransform = new TransformInternal.Component
-                {
-                    Location = (transformToSend.Position - worker.Origin).ToImprobableLocation(),
-                    Rotation = transformToSend.Orientation.ToImprobableQuaternion(),
-                    Velocity = transformToSend.Velocity.ToImprobableVelocity(),
-                    PhysicsTick = transform.PhysicsTick + data.TicksSinceLastUpdate[i].NumberOfTicks,
-                    TicksPerSecond = tickRate.PhysicsTicksPerRealSecond,
-                    DirtyBit = true
-                };
+                var currentTransform = new TransformInternal.Component();
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Why the change away from object initializer syntax?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 14]
```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
...
+<# for (var i = 0; i < (fieldDetailsList.Count / 8) + 1; i++) { #>
+                    case <#= i #>:
+                        return (dirtyBits<#= i #> & (0x1 << propertyIndex % 8)) != 0x0;
+<# } #>
+                    default:
+                        throw new ArgumentException("propertyIndex argument out of range.");
+                }
+            }
+
+            // like the IsDirty() method above, the propertyIndex arguments starts counting from 0.
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Nit: capital "Like"

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 15]
```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
...
+            private byte dirtyBits<#= i #>;
+<# } #>
+
+            public bool IsDirty()
+            {
+                return isDirty;
+            }
+
+            /*
+            The propertyIndex arguments starts counting from 0. It depends on the order of which you defined
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: "The propertyIndex arguments starts counting from 0. It depends on the order of which you defined your component properties in a schema component but is not the schema field number itself."
->
"The propertyIndex arguments counts up from 0, in the order defined in the schema component. It is *not* the schema field number itself."

"e.g." -> "For example,"

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 16]
```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
...
+            In that case, val_a uses propertyIndex 0 and val_b uses propertyIndex 1 in this method.
+            */
+            public bool IsDirty(int propertyIndex)
+            {
+                if (propertyIndex < 0 || propertyIndex >= <#= fieldDetailsList.Count #>)
+                {
+                    throw new ArgumentException("propertyIndex argument out of range.");
+                }
+
+                var byteBatch = propertyIndex / 8;
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: This looks suspicious - a comment explaining why it works would be good.
Doubly-suspicious is the need for the `default` case below. If the range checking logic is sound, that should be unnecessary.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 17]
```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
...
+
+            // Bit masks for tracking which component properties were changed locally and need to be synced.
+            // Each byte tracks 8 component properties.
+<# for (var i = 0; i < (fieldDetailsList.Count / 8) + 1; i++) { #>
+            private byte dirtyBits<#= i #>;
+<# } #>
+
+            public bool IsDirty()
+            {
+                return isDirty;
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: I'd recommend removing this flag, as it's redundant.
`return dirtyBits0 || dirtyBits1 || ...;` is just as comprehensible.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 18]
```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
...
+                        break;
+<# } #>
+                    default:
+                        throw new ArgumentException("propertyIndex argument out of range.");
+                }
+
+                isDirty = true;
+            }
+
+            public void MarkNotDirty()
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: `MarkDataDirty`, `MarkDataClean`? I shy away from including negatives in names.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 13 above)
Comment (jared-improbable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Quite, but object initialization syntax doesn't bypass the `set` methods, does it?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 14 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Fixed! :)

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 17 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Yeah, this "optimisation" was probably unnecessary. Changed to your suggestion.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 15 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Fixed!

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 13 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Uhh, that very nice. Changed back to old syntax.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 16 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Removed the default case. That code should indeed never be called. Added a comment to clarify the approach.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs [hunk 19]
```diff
@@ -51,12 +51,13 @@ protected override void OnUpdate()
...
-                    DirtyBit = true
+                    TicksPerSecond = tickRate.PhysicsTicksPerRealSecond
                 };
 
-                if (!(TransformUtils.HasChanged(currentTransform.Location, transform.Location) ||
-                    TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation)))
+                var locationHasChanged = TransformUtils.HasChanged(currentTransform.Location, transform.Location);
+                var rotationHasChanged = TransformUtils.HasChanged(currentTransform.Rotation, transform.Rotation);
+
+                if (!locationHasChanged && !rotationHasChanged)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Please revert this.

**Comment by samiwh** [comment]
I think this is a good exmaple where a small amount of git magic would make this much easier to read. There are ~2000 changed lines, most of which are generated code. But if we knew which commits to look in it wouldn't take so long to find which files we are meant to be looking at.

File: workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs (hunk 19 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs]: Reverted!

**Comment by JonasImprobable** [comment]
@samiwh sorry about that. For this R it's sufficient to just look at the aggregate changes made to the following files:
test-project/Assets/EditmodeTests/Components/BlittableComponentTests.cs
test-project/Assets/EditmodeTests/GameObjectRepresentation/Readers/WriterSendUpdateTests.cs
workers/unity/Packages/com.improbable.gdk.transformsynchronization/Systems/RateLimitedTransformSendSystem.cs
workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt
workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt
workers/unity/Packages/com.improbable.gdk.core/Components/ISpatialComponentData.cs

THe rest is generated code for the test project.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 20]
```diff
@@ -35,7 +107,7 @@ namespace <#= qualifiedNamespace #>
                 get => <#= fieldDetails.CamelCaseName #>;
                 set
                 {
-                    DirtyBit = true;
+                    MarkDataDirty(<#= i #>);
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Observation (no change needed) - this will be unstable with reordering two components in schema, in such a way that the schema doesn't change (field ids are the same).

I don't think this is a problem or something that we can do anything about in this PR. But might be something we need to be aware of if we are able to do something with serialising components.

File: test-project/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveSingular.cs [hunk 21]
```diff
@@ -63,7 +140,7 @@ public int Field4
                 get => field4;
                 set
                 {
-                    DirtyBit = true;
+                    MarkDataDirty(3);
```
Comment (samiwh) [inline test-project/Assets/Generated/Source/improbable/gdk/tests/ExhaustiveSingular.cs]: What a troll. Had to check the schema file.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 20 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: We could let `IsDataDirty()` and `MarkDataDirty()` use the schema field number of a property. In that case, we would scan for all field numbers and potentially create more byte bitmasks than necessary in case the field numbers have gaps but it would make the generated code more invariant.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 22]
```diff
@@ -24,8 +24,80 @@ namespace <#= qualifiedNamespace #>
...
+                bool val_a = 1;
+                bool val_b = 3;
+            }
+            In that case, val_a corresponds to propertyIndex 0 and val_b corresponds to propertyIndex 1 in this method.
+            */
+            public bool IsDataDirty(int propertyIndex)
+            {
+                if (propertyIndex < 0 || propertyIndex >= <#= fieldDetailsList.Count #>)
+                {
+                    throw new ArgumentException("propertyIndex argument out of range.");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Quotes around propertyIndex

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Maybe say what the provided index was and what the acceptable range is? Could help in debugging

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 23]
```diff
@@ -24,8 +24,80 @@ namespace <#= qualifiedNamespace #>
...
+                var dirtyBitsByteIndex = propertyIndex / 8;
+                switch (dirtyBitsByteIndex)
+                {
+<# for (var i = 0; i < (fieldDetailsList.Count / 8) + 1; i++) { #>
+                    case <#= i #>:
+                        return (dirtyBits<#= i #> & (0x1 << propertyIndex % 8)) != 0x0;
+<# } #>
+                }
+
+                return false;
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: This can be a default case in the switch

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt [hunk 24]
```diff
@@ -24,8 +24,80 @@ namespace <#= qualifiedNamespace #>
...
+            }
+
+            // Like the IsDataDirty() method above, the propertyIndex arguments starts counting from 0.
+            public void MarkDataDirty(int propertyIndex)
+            {
+                if (propertyIndex < 0 || propertyIndex >= <#= fieldDetailsList.Count #>)
+                {
+                    throw new ArgumentException("propertyIndex argument out of range.");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Similar comments about the error message here as above

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 22 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Done and done. :)

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 23 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: I was asked to remove the default case. I'm happy to put it back in if @jared-improbable is happy with that?

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 24 above)
Comment (JonasImprobable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: Done.

File: workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt (hunk 23 above)
Comment (jared-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt]: My objection to the `default` case was the fact that it could never be run. Hoisting the `return` up into it is fine.
